### PR TITLE
Add a Stimulus controller to handle back end tooltips 

### LINKED ISSUE #7445: Hovering over icons in backend does not show the description
### Affected version(s)

5.4.0

### Description

On some or most pages in the backend, there is no description shown when hovering with the mouse over an icon. Some example pages where this happens (not) are "Seiten", "Artikel", "Themes", "Templates", "Einstellungen". After refreshing the page (Strg+R) it works as expected. On other pages, like e.g. "Dateien", it works without refreshing the page. Interestingly, it works with Google Chrome initially on page "Seiten" sometimes, but refreshing the window, clicking through some other pages in backend, it stops working again.

Tested with Firefox (129.0.1 private window), Google Chrome 127.0.6533.120 (Offizieller Build) (64-Bit).

### PR DESCRIPTION
**Pull request by zoglo**
_And yet another stimulus controller_

* Potential fix for #7445

This rewrites the tooltip.js to a Stimulus controller.
Using the data-controller on the `body` instead of the document makes sure that the tooltip container is initialized on render (otherwise we'd have to add event listeners).

This also introduces some refactoring for the configurations to make it more readable.

### CODE REVIEW

File: core-bundle/contao/templates/backend/be_main.html5 [hunk 1]
```diff
@@ -37,7 +37,7 @@
   <?php /* If the following hash changes, Turbo will reload the entire page instead of merging contents. */ ?>
   <meta data-turbo-track="reload" content="<?= md5($this->stylesheets."\0".$this->javascripts); ?>">
 </head>
-<body id="top" class="be_main<?php if ($this->isPopup): ?> popup<?php endif; ?>"<?= $this->attributes ?>>
+<body id="top" class="be_main<?php if ($this->isPopup): ?> popup<?php endif; ?>"<?= $this->attributes ?> data-controller="contao--tooltip">
```
Comment (fritzmg) [inline core-bundle/contao/templates/backend/be_main.html5:40]: I think for consistency we should register this controller on `<html>`, just like the other(s).

Comment (fritzmg) [inline core-bundle/contao/templates/backend/be_main.html5:40]: I think for consistency we should register this controller on `<html>`, just like the other(s).

Comment (zoglo) [reply core-bundle/contao/templates/backend/be_main.html5:40]: This won't remove the stimulus controller when using Turbo but maybe the already added MutationObserver within this controller will handle new elements - Will try it out.

**Comment by zoglo** [comment]
What about deprecating tool tips like that in Contao 5 and using real targets in Contao 6 🤔?

**Comment by fritzmg** [comment]
Not sure that's feasible? Having to add `data-tips-target` to all these places does not seems sensible to me.

**Comment by zoglo** [comment]
Maybe it's not that many, will have to elaborate that first, however that's not for this current case as it's something that could be like that in C6 🤔

And we did something like that for the scroll offset controller 🙃 ...

**Comment by m-vo** [comment]
> Not sure that's feasible? Having to add `data-tips-target` to all these places does not seems sensible to me.

IMHO a single controller with targets for each tooltip seems the most reasonable implementation to me. Migrating classes to data-attributes could be done bit by bit with something like https://github.com/contao/contao/pull/7408/commits/1fec01f53c6b4751f69c7fc530675aca2c8191aa#diff-ed81209a74bff6f3ac75eb4a064808c9dad25b40923bb9f62f46dd89938081fdR25-R41, no?

**Comment by fritzmg** [comment]
From a dev perspective I find it inconvenient, that I then always have to add the data attribute when creating custom buttons etc. instead of the tooltip working automatically across the back end.

**Comment by m-vo** [comment]
IMHO anything outputting HTML (like buttons for instance) should be done via a template anyways, so we could provide a generic one where things are already set up. Just thinking out loud. 🙂

**Comment by zoglo** [comment]
> IMHO anything outputting HTML (like buttons for instance) should be done via a template anyways, so we could provide a generic one where things are already set up. Just thinking out loud.

Loud enough about what I thought as well.

It's something to elaborate in the first place if it's feasible to do so and if it's simply added and generated within DC_Table and the likes.

Once we have the possibility to opt this out:

There is still the feature request about adding a context menu, by that time half of these titles will still be needed but we wouldn't want the titles to be shown by a tooltip.

**Comment by zoglo** [comment]
Kinda wanna discuss if we should do either of these:

* leave it as it is and let the controller on the `<body>` run the querySelectorAll and attach the events
* move it to `<html>`, let the added MutationObserver handle newly added nodes and attach the events when Turbo is in progress and just replaced the whole `<body>`

or:
* follow https://github.com/contao/contao/pull/7458#issuecomment-2295063002 and add the attributes to the specific targets (meaning we'll add targets to the controller and not the controller handling everything)
* Closing this PR in favor of just leaving the solution as it is and checking whether a tooltip document exists on body.. and if not, we just create it (simply creating it for the case of turbo prefetched sites)

As mentioned in https://github.com/contao/contao/pull/7458#issuecomment-2295201644, I'd prefer to make it an Opt-In with a `data-target` for Contao 6... worst case for anyone not using this in their custom implementation... the browser-native implementation would show the title :D

**Comment by fritzmg** [comment]
> * leave it as it is and let the controller on the `<body>` run the querySelectorAll and attach the events
> 
> * move it to `<html>`, let the added MutationObserver handle newly added nodes and attach the events when Turbo is in progress and just replaced the whole `<body>`

We need to leave it as it is. If we move the controller to `<html>`, we would instead always need to check whether the `<div>` still exists, as it will be removed when Turbo replaces the `<body>`. Thus it is better to rely on the controller's lifecycle instead to always add and remove it appropriately.

> * follow https://github.com/contao/contao/pull/7458#issuecomment-2295063002 and add the attributes to the specific targets (meaning we'll add targets to the controller and not the controller handling everything)

Imho we should keep it as it is for a bugfix release.

> * Closing this PR in favor of just leaving the solution as it is and checking whether a tooltip document exists on body.. and if not, we just create it (simply creating it for the case of turbo prefetched sites)

If you find a way to fix the original JavaScript that would be fine with me too. However, you already did the work for a Stimulus controller, so why not just use that.

**Review by fritzmg — DISMISSED** [review]

File: core-bundle/contao/templates/backend/be_main.html5 (hunk 1 above)
Comment (zoglo) [reply core-bundle/contao/templates/backend/be_main.html5:40]: See: https://github.com/contao/contao/pull/7458#issuecomment-2298259378

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @zoglo.

**Pull request merged**